tests/performance/timing_matrix_free_kokkos.cc

```diff
@@ -0,0 +1,342 @@
+  {
+    matrix_free.initialize_dof_vector(vec);
+  }
+
+  void
+  vmult(VectorType &dst, const VectorType &src) const
+  {
+    LaplaceOperatorLocal<dim, fe_degree, Number> local_operator;
+    matrix_free.cell_loop(local_operator, src, dst);
+  }
+
+private:
+  CUDAWrappers::MatrixFree<dim, Number> matrix_free;
+};
+
+
+
+template <int dim, typename T>
+class AnalyticalFunction : public Function<dim, T>
+{
+public:
+  virtual T
+  value(const Point<dim, T> &p, const unsigned int component = 0) const override
+  {
+    (void)component;
+
+    double temp = 0.0;
+
+    for (unsigned int d = 0; d < dim; ++d)
+      temp += std::sin(p[d]);
+
+    return temp;
+  }
+};
+
+
+
+template <unsigned int dim, const int degree, typename MemorySpace>
+std::vector<double>
+run(const unsigned int n_refinements)
+{
+  ConvergenceTable table;
+
+  const MPI_Comm comm = MPI_COMM_WORLD;
+
+  using Number     = double;
+  using VectorType = LinearAlgebra::distributed::Vector<Number, MemorySpace>;
+
+  const unsigned n_repetitions = 100;
+
+  parallel::distributed::Triangulation<dim> tria(comm);
+
+  GridGenerator::hyper_cube(tria);
+  tria.refine_global(n_refinements);
+
+  table.add_value("n_levels", tria.n_global_levels());
+  table.add_value("degree", degree);
+
+  table.add_value("n_cells", tria.n_global_active_cells());
+
+  const MappingQ1<dim> mapping;
+  const FE_Q<dim>      fe(degree);
+  const QGauss<1>      quadrature(degree + 1);
+
+  DoFHandler<dim> dof_handler(tria);
+  dof_handler.distribute_dofs(fe);
+
+  table.add_value("n_dofs", dof_handler.n_dofs());
+
+  AffineConstraints<Number> constraints;
+
+  std::chrono::time_point<std::chrono::system_clock> temp =
+    std::chrono::system_clock::now();
+  LaplaceOperator<dim, degree, Number, MemorySpace> laplace_operator(
+    mapping, dof_handler, constraints, quadrature);
+
+  const double dt_setup = std::chrono::duration_cast<std::chrono::nanoseconds>(
+                            std::chrono::system_clock::now() - temp)
+                            .count() /
+                          1e9;
+
+  VectorType src, dst;
+
+  laplace_operator.initialize_dof_vector(src);
+  laplace_operator.initialize_dof_vector(dst);
+
+  {
+    LinearAlgebra::distributed::Vector<Number> src_host(src.get_partitioner());
+
+    VectorTools::interpolate(dof_handler,
+                             AnalyticalFunction<dim, Number>(),
+                             src_host);
+
+    LinearAlgebra::ReadWriteVector<Number> rw_vector(
+      src.get_partitioner()->locally_owned_range());
+    rw_vector.import(src_host, VectorOperation::insert);
+    src.import(rw_vector, VectorOperation::insert);
+
+    dst = 0.0;
+  }
+
+  double dt_vmult = 0;
+
+  for (unsigned int i = 0; i < n_repetitions; ++i)
+    {
+      MPI_Barrier(MPI_COMM_WORLD);
+
+      std::chrono::time_point<std::chrono::system_clock> temp =
+        std::chrono::system_clock::now();
+
+      laplace_operator.vmult(dst, src);
+
+      MPI_Barrier(MPI_COMM_WORLD);
```

Comment:
I would time the sum over all vmults without barriers inside.

Reply:
Indeed, there should not be barriers around single `vmult` operations. It is more realistic to test for many `vmult` operations in sequence (e.g., explicit time stepping, simple relaxations, Chebyshev iterations). Of course, we will not often have 100 (or 500, whatever useful count) operations using the exact same communication pattern, but it's a pretty good approximation.